### DISCUSSION TITLE
Is there a concise way of having a thread yield on a signal from another thread without "merging" the function with that thread?

### DISCUSSION THREAD

**Discussion by KnightNine** [godotengine/godot-proposals#5562] [Engine Core]
It has come to my attention that yielding on anything that is within another thread will add the function call that is yielding to that other thread. i.e.:

```gdscript
var thread = Thread.new()
signal thread_done()
func thread_func():
	emit_signal("thread_done")

func _ready():
	
	thread.start(self,"thread_func")
	yield(self,"thread_done")
	
	#(for demonstration purposes) this will throw an error because this _ready() call is now part of the thread
	thread.wait_to_finish()
```

 I am able to get back to the main thread by calling `yield(get_tree(), "idle_frame")` but how do I get back into another thread?
 like if I wanted to call something from the main thread and yield upon it for threadsafety reasons, how would I get my function back into the thread I originally called it from?
 
here is an example similar to why I need this functionality 
 
 ```gdscript
 signal thread_done()
func thread_func():
	
	#upon yeilding for the result of this function the thread_func() is no longer running from within the thread
	var is_colliding = yield(check_if_colliding(),"completed")
	
	#my thread is meant to continue running other things in the background after this call as to not burden the main thread. 
	#But i am unable to yield on anything that was called from the original thread, so what am i to do?
	
	call_deferred("emit_signal","thread_done")

func check_if_colliding():
	
	var is_colliding = false
	print("start")
	
	var col_shape = CollisionShape.new()
	var cylinder = CylinderShape.new()
	#add to world
	Global_Variables_3D.world.add_child(col_shape)
	col_shape.shape = cylinder
	
	#do whatever positioning with the col_shape and cylinder here...
	
	#wait until col_shape is added to the world (this also merges this with the main thread)
	yield(col_shape,"ready")
	
	var collider_shape_params = PhysicsShapeQueryParameters.new()
	var exclude_arr = []
	collider_shape_params.set_shape(col_shape.shape)
	collider_shape_params.set_collide_with_areas(false)
	collider_shape_params.set_collide_with_bodies(true)
	collider_shape_params.set_exclude(exclude_arr)
	
	#need to transform the collider shape params itself:
	collider_shape_params.transform = col_shape.transform
	
	#merge with the physics thread in order to safely call the collision functions
	yield(get_tree(),"physics_frame")
	
	var phys_dir_space = get_world().direct_space_state
	
	#need to transform the collider shape params itself:
	collider_shape_params.transform = col_shape.transform
	
	var results = phys_dir_space.collide_shape(collider_shape_params,1)
	if results.size() != 0:
		is_colliding = true
	
	
	#now how do i get this function to merge back to the thread i called it from?
	
	
	return is_colliding

func _ready():
	
	thread.start(self,"thread_func")
	yield(self,"thread_done")
	
	thread.wait_to_finish()
```

Here would be my solution to this which would re-merge the function to a thread but I'm not sure if there's a better way of doing what I'm trying to do:

```gdscript
signal _thread_hook()
func _thread_hook():
	emit_signal("_thread_hook")
	
func _create_hook():
	thread.wait_to_finish()
	thread.start(self,"_thread_hook")

#this needs to be called as yield(reattach_to_thread(),"completed") in order to work
func reattach_to_thread():
	#the thread is created to emit a signal
	_create_hook()
	#this yield merges/hooks the function back into the thread
	yield(self,"thread_hook")
	print("hooked")
```

**Comment by KnightNine** [comment]
what would be nice if there was something like "preserve_thread_yield()" that would yield without leaving the thread

i just realized that :
```gdscript
thread.wait_to_finish()
thread.start(self,"_thread_hook")
```
Is stuttery and slow AF so it's not even a viable solution when I'm calling this multiple times. (likely as this bug report states: https://github.com/godotengine/godot/issues/24869)

I'm pretty stuck here.

**Comment by KnightNine** [comment]
I also tried to create my own "process" loop within a thread that I could use as the source for a signal, and that didn't work since the thread won't loop forever. 

```gdscript
var thread_active = true
signal thread_frame()
var i = 0
func thread_process():
	
	while thread_active:
		#this loop stops at around 2000, there's probably something stopping threads from looping forever
		print("t:"+str(i))
		emit_signal("thread_frame")
		i+=1
		

```

I think that's my last idea for solving this issue within gdscript...

Update:
I think I actually managed to fix this by adding a semaphore to the loop, but I'll need to implement it into my project to 100% confirm that it runs without issue. 

```gdscript
var semaphore = Semaphore.new()
var thread_active = true
signal thread_frame()
#in the case where the yield() is called after emit_signal("thread_frame"), this is needed to ensure that eventually the yield catches the signal
var yielding_on_thread_process = false


func thread_process():
	
	while thread_active:
		
		#if reattach_to_thread is still yielding, keep emitting signals until this is false
		if !yielding_on_thread_process:
			semaphore.wait()
		#this fixes the deadlocking and error issues (maybe make the delay a bit larger than 1 usec just in case):
		OS.delay_usec(1)
		emit_signal("thread_frame")
		
		
#this needs to be called as yield(reattach_to_thread(),"completed") in order to work
func reattach_to_thread(printer):
	yielding_on_thread_process = true
	semaphore.post()
	yield(self,"thread_frame")
	yielding_on_thread_process = false
	print("this call should now be within thread" + str(printer) + "------------------")
#	thread.wait_to_finish()

func _ready():
	thread.start(self,"thread_process")
	
	yield(reattach_to_thread("print test in thread"+str(0)),"completed")
	
	yield(get_tree(),"physics_frame")
	
	yield(reattach_to_thread("print test in thread"+str(1)),"completed")
	print("done")
	#this will throw error
	thread.wait_to_finish()
```

Update 2:
After some testing, It does work without issue but running it 700 times is extremely slow, taking over 9 seconds.
I wonder if there's a way to speed this up...

`semaphore.post()` can get deadlocked if it posts before the semaphore waits, not sure how to fix that yet

Update 3:
 added `OS.delay_usec(1)` due to bug https://github.com/godotengine/godot/issues/67785

**Comment by KnightNine** [comment]
Turns out I'm retarded and can't read, it was actually yielding on the physics frame which was causing the slowness in my project.
so the only real issue was the deadlocking:
![i can't read](https://user-images.githubusercontent.com/34152090/195198479-9eecb6d8-6385-4080-abb5-27b9317f3d51.PNG)

I also figured out that the deadlocking wasn't deadlocking but some strange `emit_signal: Error calling method from signal 'thread_frame': 'GDFunctionState::': Method not found..` error that I can't make out and that happens at random. But I managed to circumvent it by adding a counter at the beginning of the `while thread_active:` loop, idk why it fixes the issue but it does . 
(EDIT: [i think i figured out why](https://github.com/godotengine/godot/issues/67785))

**Comment by dave2** [comment]
I've just into this issue myself, I had a tween running in a thread, so naturally wrote

`await create_tween().tween_property(..).finished`

and this caused the execution to fall out of the thread. If I replaced all the cases of await with spinloops and delays it works fine, albeit stuttery (presumably because we're calling `OS.delay_msec()` a lot in my spinloops.
